src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -796,68 +548,36 @@ private void InterceptException(Action action)
         /// Sets a new writer state; verifies that the transition from the current state into new state is valid.
         /// </summary>
         /// <param name="newState">The writer state to transition into.</param>
-        private void SetState(BatchWriterState newState)
-        {
-            this.InterceptException(() => this.ValidateTransition(newState));
+        protected abstract void SetState(BatchWriterState newState);
 
-            switch (newState)
+        /// <summary>
+        /// Wrapper method to validate state transition with optional customized validation.
+        /// </summary>
+        /// <param name="newState">Teh new writer state to transition into.</param>
+        /// <param name="customizedValidationAction">Optional validation action.</param>
+        internal void ValidateTransition(BatchWriterState newState, Action customizedValidationAction)
+        {
+            if (customizedValidationAction != null)

```

Comment:
>customizedValidationAction [](start = 16, length = 26)

ValidateTransition should only validate that the state transition is valid, which should be independent of any implementation so it shouldn't require passing in a customValdiationAction.